Stop using bold text for input and answers

jayschwa opened the pull request:
**This pull request turns off bold text everywhere and changes it to "normal" text.**

**Julia should respect the user's font settings; if they really wanted to see bold text everywhere, they presumably would have adjusted it in their terminal's settings. I don't think there's a compelling reason for Julia not to toe the line here. For the existing Julia users who like the good old days, bold text can still be enabled via the `JULIA_[INPUT|ANSWER]_COLOR=bold` environment variables.**

Bold styling should be used conservatively to **highlight** information. Switching the baseline to normal text opens the door for more intelligent use of bold in the future. :rocket:

StefanKarpinski commented:
Normal printed output is non-bold. I find that the difference helps make it clearer what's going on. We already make colored responses opt-in. At this rate, we're headed towards the git situation where 99% of users don't know how to make it look colorful or nice.

andreasnoack commented:
+1. I've had this locally for a while. I agree that bold should only be used for emphasizing pieces of text.

ssfrr commented:
This would also resolve #8096 (colors would show up correctly for people using the Solarized colorscheme). Note that (as I mention in that issue) I think that this is basically Solarized's fault for changing the meaning of bold text, but it is a relatively popular theme, and it seems weird to use bold unnecessarily.

@StefanKarpinski isn't the color by itself enough to differentiate from normal printed text? Maybe colored text could print in non-bold color and if color is turned off all colored text just gets bolded instead?

andreasnoack commented:
I just got annoyed by the bold fonts again. It really makes the text harder to read and I agree with @jayschwa that it is wrong to use bold here so I'm in favor of merging this.

jayschwa commented:
Rebased from master and resolved merge conflict.

andreasnoack commented:
@jayschwa Thanks. I'll merge tomorrow unless someone objects.

blakejohnson commented:
@andreasnoack I didn't sense a consensus on this one, yet.

hayd commented:
Being able to disambiguate at the repl:
```
f() = println("\"hello\"")
g() = "hello"
```
seems useful. I don't care about `julia> INPUT` but output should be emphasized.

andreasnoack commented:
I agree that it should be possible to disambiguate `f` and `g` in your example, but I don't think bold is the right choice. IPython, R and Matlab print `Out[_x_]`, `[_linenumber_]` and `ans = ` respectively before output. We could consider printing something similar or alternatively we could use a different color for output.

mmoh commented:
Is it possible to enable bold (or a color) on windows, or is this a linux only feature? I don't have anything bolded.

jayschwa commented:
@mmoh, Julia on Windows uses non-bold text by default. Bold doesn't appear to do much in a Windows terminal. It brightens the color a bit, but doesn't change the visual weight of the characters. Here's how to play with it:
``` julia
print_with_color(:bold, "This is bold")
```

tkelman commented:
I think it should work in mintty though

andreasnoack commented:
@jayschwa Do you have a proposal for distinguishing the return from a function from what is written to STDOUT?

jayschwa commented:
Is `JULIA_ANSWER_COLOR` not sufficient?

andreasnoack commented:
I think we should set a default value. `:cyan` or `:magenta` would be fine with me.

jayschwa commented:
Using a prefix seems more in line with what other language REPLs do, but I feel like that falls outside the scope of this PR. I personally can't recall many times where differentiating between a return value and stdout was problematic for me.

StefanKarpinski commented:
I have return values as `:magenta` which looks quite nice. May even look ok on Windows. IIRC, cyan was bad on Windows.

mmoh commented:
How do you change the defaults and make them permanent?

andreasnoack commented:
Let's try with default output in magenta. Then I think we have a compromise that will eliminate the bold letters and still disambiguate printed output from return values. @jayschwa, if you update then I'll merge.

jakebolewski commented:
NO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

jayschwa commented:
> NO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

I concur, I think magenta will be more offensive to my eyes than bold :see_no_evil:

I'm skeptical that there is widespread confusion between prints and REPL output. `JULIA_ANSWER_COLOR` is available for those who want to tweak it.

andreasnoack commented:
@StefanKarpinski Color seems to be a sensitive issue. Should we try without and see if it causes confusion that return values in the REPL and printed output are identical? I like the magenta output idea, but my main priority is to get rid of the bold letters.

StefanKarpinski commented:
Apparently the worst bikeshed is the original: what color to paint the bike shed.

StefanKarpinski commented:
I worry that this now defaults to giving people the least information and the least "fancy" setup. This is kind of like how git defaults to not using colors. As a result, I've seen people who've used git for years with no colors. I turn the option on for them and suddenly the whole experience is changed. I don't want Julia to end up like this. Even if you don't like the default output, at least the use of colors and bold lets you know that colors and bold are an option and something that you can configure.

jakebolewski commented:
Why does bold == fancy?  It seems like a subjective opinion. I don't like that you take away bold as an option to emphasize particular output.  So it seems like you can add _more_ information to output when bold by default is turned off.

hayd commented:
If it's easy and well documented to turn off the bold, what's the problem?

\<bikeshed> I'm not sure magenta is a good choice either. \</bikeshed>

jayschwa commented:
> If it's easy and well documented to turn off the bold, what's the problem?

The problem is outlined in the original post. Matching the terminal's font shouldn't require explicit action, it should happen by default.

hayd commented:
IMO it's slightly disingenuous to say this could be a terminal setting (rather than a julia setting). Being bold is still "matching the terminal font", having *everything* be bold  (and I don't think anyone is advocating that) is a very different behavior than having *output* be bold... so this *has* to be a julia setting?

That said, I think color, or bold + color, might well be a better default than bold + white.

StefanKarpinski commented:
I'm fine with getting rid of bold, but only if we make the output colored by default – we need to give some indication that color is possible, even if people then want to change it. Do people feel that cyan is less repugnant than magenta? It looks pretty similar (but different) from white output.

mmoh commented:
+1 magenta looks good, cyan looks terrible on default colorscheme on windows

andreasnoack commented:
Let's try with the magenta default. We'll then get rid of the bold fonts while still having a visible difference between REPL output and print/show output. Users who don't like colors can then set `JULIA_ANSWER_COLOR=white`. This should be contrasted with the present situation in which you cannot get colored non-bold answers.

@jayschwa I think this is the best compromise if you want the bold fonts to go away.

jayschwa commented:
My dev machine is out of order while I do some hardware upgrades. Someone else feel free to fork my branch and open a new PR. :fork_and_knife:

cormullion commented:
I searched and found this issue. I don't know whether there's something wrong with my machine or the fonts, or with the Julia REPL.

Here's a screen shot:

![screen shot 2016-02-07 at 12 12 25](https://cloud.githubusercontent.com/assets/52289/12872207/17632ad0-cd94-11e5-9a45-acac0e481854.png)

I'm not a fan of the way that the banner text and the prompts are bolded - it emphasizes the wrong things. I don't mind that user input is bolded, but the return values are bolded too, as well as the prompt, so - when everything is bolded, nothing's bolded... :)  Although it's true that `println`s output isn't bolded...

I like the Julia colors. I just think the bolding looks clumsy.

cormullion commented:
Is there an option to not have the bold text I get by default? (I assume this PR isn't going anywhere...)

nalimilan commented:
Would it be possible to keep the bold only for code, and remove it for all output (i.e. result of `print`/`show` *and* standard printing of values)? That would be a legitimate use of bold IMHO. The current inconsistency between different types of printing is confusing for me.

KristofferC commented:
+1 for this. The fact is that in many terminals bold does not only mean bold but bold **and** use the highlighted version. This means that the colors we are getting are not really the colors that `:yellow` and company are supposed to represent.

As an example:

![image](https://cloud.githubusercontent.com/assets/1282691/18472843/0ce80490-79ba-11e6-883a-fdeb20553b98.png)

KristofferC commented:
I guess the one problem here is that with a PR like this there is currently no way for someone who wants to use bold colored input text. The API for choosing the text style is quite restricted because it uses an ENV var (a string) so it doesn't compose so nicely (like `input_color!(foreground = :red, bold = true)` would have). I've been scratching my head a bit and I see two simple solutions but both can be considered a bit ugly.

* Introduce a complementary ENV for each customization variable that represents if bold is active or not. So `ENV["JULIA_INPUT_BOLD"] = true` etc.

* Add bold versions to all different colors `ENV["JULIA_INPUT_COLOR"] = :bold_red` etc.

Both of these feel a bit awkward though.

StefanKarpinski commented:
@KristofferC – since you've had the REPL bug this release, do you want to yea or nay this? If yay, push it forward in whatever manner you see fit.

KristofferC commented:
Some random thoughts and experiments here. Firstly, I support not using bold for input text. However, that makes bold answer color feel a bit "too strong". So the natural continuation is to stop using bold answer color. That leads to the problem @hayd mentioned, that displayed values and printed strings are not distinguishable by default. Since the majority of users are likely to run with default configuration I think it is important to have a default setting that distinguishes these. So I tried @andreasnoack's idea with an output indicator:

![image](https://cloud.githubusercontent.com/assets/1282691/21663430/6e0a7b48-d2e0-11e6-87a0-65b2c084c934.png)

I kind of like this except it breaks output for objects that want to align text on the first line:

![image](https://cloud.githubusercontent.com/assets/1282691/21663568/576d00c6-d2e1-11e6-8ed2-be5f89d2186d.png)

I looked a bit at IPython and they somehow figure out that the matrix wants to be aligned and prints an initial new line that is not visible in the standard python REPL:

![image](https://cloud.githubusercontent.com/assets/1282691/21663711/2da750d8-d2e2-11e6-83e1-c55599a6000d.png)

I'm not sure what is best here. There are many opinions and it is hard to find something that is obviously the right thing to do. Personally, in order to print something in bold, there should be a good reason for it and you want that text to stick out or carry extra weight. To me, applying it to all input and output fails that criteria. 

Happy to discuss this more.

JeffBezanson commented:
I agree with the idea of this issue. I just tried making only answers and `julia>` bold and it seems fine to me. Somehow making the prompt bold as well might avoid answers appearing to stand out too much.

jayschwa commented:
It seems like a lot of thought is being put into making prints visually different from REPL answers. Are people getting the two confused an actual problem? To me, it seems like it should be relatively easy to differentiate between the two just based on their order. The answer is at the end. If there are any prints, they will be sandwiched between the input and the answer. Am I missing something?

JeffBezanson commented:
You might be right. Let's just rebase this and try it out.

jayschwa committed:
Stop using bold text for input and answers

Some hardcoded ANSI codes were also replaced with `text_colors`.

jayschwa commented:
I have rebased the PR off master and replaced some hardcoded ANSI codes with `text_colors`.

@KristofferC, I did not remove the bold from [`input_color()` and `answer_color()`](https://github.com/JuliaLang/julia/blob/master/base/client.jl#L84-L86). As far as I can tell, they don't actually affect anything. Might be worth investigating!

JeffBezanson commented:
:+1:

KristofferC commented on base/REPL.jl:
```diff
@@ -958,7 +956,7 @@ type StreamREPL <: AbstractREPL
     StreamREPL(stream,pc,ic,ac) = new(stream,pc,ic,ac,false)
 end
-StreamREPL(stream::IO) = StreamREPL(stream, julia_green, Base.input_color(), Base.answer_color())
+StreamREPL(stream::IO) = StreamREPL(stream, Base.text_color[:green], Base.input_color(), Base.answer_color())
```
This should probably be `light_green` if we want to keep the same color as before.

tkelman replied:
it should also be `text_colors`, which tells me this line isn't hit by tests at all

KristofferC replied:
When is this even used?

tkelman replied:
remote repl maybe? untested stuff that probably only Keno uses?

jayschwa replied:
Re-pushed with the typo fixed. `:green` was left as-is because that matches what it was before. The only difference is that it's no longer contains with the code for bold, but that was redundant since bold is prefixed later on by `Prompt`.

KristofferC replied:
No, because on most terminals (urxvt, gnome-terminal, cmd etc) bold makes text bold AND also makes the color brighter (the light versions of the color).

![image](https://cloud.githubusercontent.com/assets/1282691/21745184/9f19604a-d525-11e6-9183-95d007262041.png)

The purpose here is to strip the bold part and not change the color, hence it should be `:light_green`.

jayschwa replied:
The bold is still being applied, just somewhere else, so the color isn't actually changing (nor the bold). I see what you mean though, so I will updated it. It probably makes sense to do the same in the banner too, right?

KristofferC replied:
Oh, if the boldness is being applied then it doesn't really matter. I remember now that I had to add that here: https://github.com/JuliaLang/julia/pull/18628/files#diff-8dc8200a114faecb709b7b96e96fc0cfR632.

jayschwa replied:
I'm going to leave the prompts bold. Changing the color to `light_green` still might make sense because it decouples the color from its boldness. However, do the light variants work reliably across platforms?

KristofferC replied:
Everywhere I have tested, yes. Errors are currently being printed in `light_red` for example.

KristofferC commented:
This neeeds a NEWS entry preferably with a link how to change (and get back the bold) for the input / ouput color.

jayschwa commented:
Pushed an update that contains a NEWS entry for the style change.

jayschwa commented:
@KristofferC, I just updated the prompt and banner colors to use light variants. In the future, if bold is turned off on any of these things, their color will be unaffected. Dots in the banner are still bold, but I did unbold the ASCII art "julia" part since I think it looks better.

Are there any Windows builds that are automatically generated? I'm interested to see how light variants work there, but don't let that hold up the PR from being merged.

KristofferC commented:
Can't you just print some hard coded ANSI to see how it looks on windows? With everything else being not bold I don't see a reason for having the prompt bold but I don't mind that much.

JeffBezanson commented:
Non-bold green is pretty dark in some terminals.

KristofferC commented:
Yes but see discussion at https://github.com/JuliaLang/julia/pull/11250#discussion_r95055707

jayschwa commented:
I just played with a nightly on Windows and convinced myself that light_<color> is working and equivalent to bold <color>. I think it's good to go.

> With everything else being not bold I don't see a reason for having the prompt bold

This PR has been lingering for a year and a half. I'd rather not increase the scope 😄.

KristofferC commented:
Ok, I'll make that PR then ;)

jayschwa committed:
Merge branch 'master' into the-goggles-do-nothing

jayschwa commented:
I resolved the conflict in NEWS, which was caused by all of the entries being wrapped to a different line length on master. I suggest using "Display the rich diff" when reviewing that file.

tkelman commented on NEWS.md:
```diff
@@ -151,32 +151,25 @@ Library improvements
+    For example, one way of doing this is adding `ENV["JULIA_INPUT_COLOR"] = :bold` and `ENV["JULIA_ANSWER_COLOR"] = :bold` to the `.juliarc.jl` file.
+    See the [manual section on customizing colors](http://docs.julialang.org/en/latest/manual/interacting-with-julia/#customizing-colors) for more information.
+
+  * The default color for info messages has been changed from blue to cyan, and for warning messages from red to yellow.
```
this should probably have its issue/pr reference added, it got lost at some point

musm commented:
It looks like all the underscore `light_` colors ended up breaking  the julia repl color setup for me on 0.6 on the bash subsystem for windows.  

E.g. on 0.5
![image](https://cloud.githubusercontent.com/assets/4319522/25144562/3c634388-243c-11e7-9673-2bae4a7a4652.png)

E.g. on 0.6

![image](https://cloud.githubusercontent.com/assets/4319522/25144578/48f1cf8e-243c-11e7-82d5-cbc76a195fcd.png)

Is it possible to remove the `light_` prefix so that things work as before ? I imagine I will not be the only one experiencing this issue.

KristofferC commented:
Yes, that should work fine.

musm commented:
thanks @KristofferC  mind submitting a PR or should I?

KristofferC commented:
Likely to happen faster if you do it.
